### ISSUE TITLE
File writes to Windows NFS share silently fail...

### ISSUE THREAD

**Issue by rtestardi** [dotnet/runtime#44546]
[AB#1285380](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1285380)
We are using .NET Core 3.1 on Linux.

When we write to a file on a Windows NFS server, writes silently fail.

The following program illustrates the behavior:

```c#
using System;
using System.IO;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            if( args.Length != 1 )
            {
                    Console.WriteLine("Usage: test <folder>\n");
                    return;
            }

            var filename = Path.Combine(args[0], "file.txt");
            File.WriteAllText(filename, $"Hello World as of {DateTime.Now}");
            Console.WriteLine($"File Content: '{File.ReadAllText(filename)}'");
        }
    }
}
```
The folder should be a path to a directory on a Windows NFS server.

Basically, what is happening is .net is asking for a shared advisory lock on the file handle, and then issuing a write on that handle.

While that may work against a Linux NFS server, it does not work against a Windows NFS Server, which chooses to enforce advisory locking.

The write into the local buffer cache succeeds, but the write to the back-end server fails.

.net also fails to look at the return value of close(2), and hence never notices anything wrong -- the program above succeeds and shows the file as having no data in it after the write to the file succeeds.

The following native C program exhibits the same issue, for comparison:

```c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/file.h>
int
main(int argc, char **argv)
{
  int fd;
  int rv;
  char buffer[22];
  char *filename;
  if (argc != 2) {
    printf("usage doit <file>\n");
    exit(1);
  }
  filename = argv[1];
  fd = openat(AT_FDCWD, filename, O_WRONLY|O_CREAT|O_CLOEXEC, 0666);
  if (fd == -1) {
    printf("openat(%s) failed %d\n", filename, errno);
    exit(1);
  }
  rv = flock(fd, LOCK_SH|LOCK_NB);
  if (rv == -1) {
    printf("flock(%s) failed %d\n", filename, errno);
    exit(1);
  }
  memset(buffer, 'a', sizeof(buffer));
  rv = write(fd, buffer, sizeof(buffer));
  if (rv != sizeof(buffer)) {
    printf("write(%s, %d) failed %d, %d\n", filename, (int)sizeof(buffer), rv, errno);
    exit(1);
  } 
  printf("write(%s, %d) succeeded %d\n", filename, (int)sizeof(buffer), rv);
  rv = close(fd);
  if (rv == -1) {
    printf("close(%s) failed %d\n", filename, errno);
    exit(1);
  }
  return 0;
}
```

Workarounds for the user include:

1. mount the NFS share with "-o nolock" to not propagate lock requests from the Windows NFS Server to Windows NLM.

This is what we are seeing inside the Linux VM, using strace:
```
openat(AT_FDCWD, "/home/hcsuser/mnt/14/second.txt", O_WRONLY|O_CREAT|O_CLOEXEC, 0666) = 20
fstat(20, {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(20, LOCK_SH|LOCK_NB)              = 0
fadvise64(20, 0, 0, POSIX_FADV_SEQUENTIAL) = 0
ftruncate(20, 0)                        = 0
lseek(20, 0, SEEK_CUR)                  = 0
write(20, "example text to write\n", 22) = 22
flock(20, LOCK_UN)                      = 0
close(20)       
```
Notice it seems (Linux) .net itself is sending the flock(SH), and according to strace the write succeeds with 22 bytes written.

(The write will not succeed inside Windows NFS Server, because it is incompatible with the lock mode!)

This is what we are seeing on the server (for a different run, but the same outcome):
```
11:32:59.3285886 AM               System     4              4456        IRP_MJ_CREATE        N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt NAME NOT FOUND    Desired Access: Maximum Allowed, Disposition: Open, Options: Write Through, Open Reparse Point, Attributes: N, ShareMode: Read, Write, Delete, AllocationSize: n/a
11:32:59.3287878 AM               System     4              4456        IRP_MJ_CREATE        N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt NAME NOT FOUND    Desired Access: Maximum Allowed, Disposition: Open, Options: Write Through, Open Reparse Point, Attributes: N, ShareMode: Read, Write, Delete, AllocationSize: n/a
11:32:59.3297571 AM               System     4              4456        IRP_MJ_CREATE        N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Desired Access: Maximum Allowed, Disposition: Create, Options: Write Through, Non-Directory File, Attributes: N, ShareMode: Read, Write, Delete, AllocationSize: 0, OpenResult: Created
11:32:59.3303417 AM               System     4              4456        IRP_MN_QUERY_INFORMATION N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryIdInformation
11:32:59.3303721 AM               System     4              4456        IRP_MJ_QUERY_SECURITY         N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Information: Owner, Group, DACL, SACL
11:32:59.3303847 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryStandardInformationFile, AllocationSize: 0, EndOfFile: 0, NumberOfLinks: 1, DeletePending: False, Directory: False
11:32:59.3304132 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryBasicInformationFile, CreationTime: 11/10/2020 11:32:59 AM, LastAccessTime: 11/10/2020 11:32:59 AM, LastWriteTime: 11/10/2020 11:32:59 AM, ChangeTime: 11/10/2020 11:32:59 AM, FileAttributes: A 0x400000
11:32:59.3304254 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryStandardInformationFile, AllocationSize: 0, EndOfFile: 0, NumberOfLinks: 1, DeletePending: False, Directory: False
11:32:59.3304437 AM               System     4              4456        IRP_MJ_QUERY_INFORMATION N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryFileInternalInformationFile, IndexNumber: 0x3000000000002791
11:32:59.3309122 AM               System     4              4456        IRP_MJ_CREATE        N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Desired Access: Maximum Allowed, Disposition: Open, Options: Write Through, Sequential Access, Non-Directory File, Open By ID, Open Reparse Point, Attributes: N, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
11:32:59.3310229 AM               System     4              4456        IRP_MN_QUERY_INFORMATION N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryIdInformation
11:32:59.3310509 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryBasicInformationFile, CreationTime: 11/10/2020 11:32:59 AM, LastAccessTime: 11/10/2020 11:32:59 AM, LastWriteTime: 11/10/2020 11:32:59 AM, ChangeTime: 11/10/2020 11:32:59 AM, FileAttributes: A 0x400000
11:32:59.3310618 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryStandardInformationFile, AllocationSize: 0, EndOfFile: 0, NumberOfLinks: 1, DeletePending: False, Directory: False
11:32:59.3310810 AM               System     4              4456        IRP_MJ_QUERY_INFORMATION N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryFileInternalInformationFile, IndexNumber: 0x3000000000002791
11:32:59.3310970 AM               System     4              4456        IRP_MJ_QUERY_SECURITY         N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Information: Owner, Group, DACL, SACL
11:32:59.3311156 AM               System     4              4456        IRP_MN_QUERY_INFORMATION N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryLinks
11:32:59.3311448 AM               System     4              4456        FASTIO_LOCK            N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Exclusive: False, Offset: 0, Length: 4,294,967,295, Fail Immediately: True
11:32:59.3326908 AM               System     4              4456        IRP_MJ_CLEANUP     N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   
11:32:59.3329113 AM               System     4              4456        IRP_MJ_SET_INFORMATION      N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: SetBasicInformationFile, CreationTime: 0, LastAccessTime: 0, LastWriteTime: 11/10/2020 11:32:59 AM, ChangeTime: 0, FileAttributes: n/a
11:32:59.3331090 AM               System     4              4456        IRP_MJ_SET_INFORMATION      N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: SetBasicInformationFile, CreationTime: 0, LastAccessTime: 0, LastWriteTime: 0, ChangeTime: 0, FileAttributes: n/a
11:32:59.3331472 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryBasicInformationFile, CreationTime: 11/10/2020 11:32:59 AM, LastAccessTime: 11/10/2020 11:32:59 AM, LastWriteTime: 11/10/2020 11:32:59 AM, ChangeTime: 11/10/2020 11:32:59 AM, FileAttributes: A 0x400000
11:32:59.3331622 AM               System     4              4456        FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryStandardInformationFile, AllocationSize: 0, EndOfFile: 0, NumberOfLinks: 1, DeletePending: False, Directory: False
11:32:59.3331922 AM               System     4              4456        IRP_MJ_QUERY_SECURITY         N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Information: Owner, Group, DACL, SACL
11:32:59.3335004 AM               System     4              4456        IRP_MJ_WRITE         N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt FILE LOCK CONFLICT  Offset: 0, Length: 22, I/O Flags: Write Through
11:32:59.3335631 AM               System     4              25472      IRP_MJ_SET_INFORMATION      N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: SetBasicInformationFile, CreationTime: 0, LastAccessTime: 0, LastWriteTime: 0, ChangeTime: 0, FileAttributes: n/a
11:32:59.3336067 AM               System     4              25472      FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryBasicInformationFile, CreationTime: 11/10/2020 11:32:59 AM, LastAccessTime: 11/10/2020 11:32:59 AM, LastWriteTime: 11/10/2020 11:32:59 AM, ChangeTime: 11/10/2020 11:32:59 AM, FileAttributes: A 0x400000
11:32:59.3336425 AM               System     4              25472      FASTIO_QUERY_INFORMATION  N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Type: QueryStandardInformationFile, AllocationSize: 0, EndOfFile: 0, NumberOfLinks: 1, DeletePending: False, Directory: False
11:32:59.3336615 AM               System     4              25472      IRP_MJ_QUERY_SECURITY         N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Information: Owner, Group, DACL, SACL
11:32:59.3339400 AM               System     4              4456        FASTIO_UNLOCK_SINGLE           N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   Offset: 0, Length: 4,294,967,295
11:32:59.3339829 AM               System     4              4456        IRP_MJ_CLEANUP     N:\hcsshares\03c059d4-eb0f-4b77-bcb8-14a200969455\datadata\threex\second.txt SUCCESS   
```
Note two potential problems:

1.	the NFS flock(SH) is mapped to FASTIO_LOCK(Exclusive: False) -- this is unfortunate, since Windows will now fail writes; whereas, NFS would not -- flock(SH) is never appropriate for a handle where writes may occur (.net issues this incorrectly)!
2.	the write that failed on the server (flushing from Linux buffer cache) is reported as succeeded on the client (write(2) into the buffer cache, and .net did not check the return value of close(2)) -- there is no data in the file.

For the Windows Server lock semantics, see:

Locking a portion of a file for shared access denies all processes write access to the specified region of the file, including the process that first locks the region. All processes can read the locked region.

https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex

Recommended fix:

1.  don't ask for lock(SH) on a handle that can be used for write, since if you do, writes will always fail -- the decision to use SH needs more information in .net than it has today (it needs share mode and access rights).
2. check the return value for close(2) and throw an exception when it fails -- this will typically be in a caller's using block, if they are trying to control handle lifetimes, which is good..

**Comment by danmosemsft** [comment]
For (2), do you have 3.1.9? I am wondering whether it is fixed by https://github.com/dotnet/coreclr/pull/28099 https://github.com/dotnet/corefx/pull/42988

**Comment by danmosemsft** [comment]
I edited your comment to add some triple back tick blocks for formatting

**Comment by rtestardi** [comment]
Thanks!  To be clear, I think there are two separate issues here:

1. writes from File.WriteAllText() are guaranteed to fail to Windows NFS Server, because flock(SH) is used when flock(EX) should have been used (on a handle that can be legally used for writes), and
2. the failed write is not noticed because write(2) succeeds and close(2)'s rv is not checked properly (for the flush status from the buffer cache to Windows NFS Server -- Steve thinks an exception might be being eaten by a higher level?).

It is possible the second was fixed by 3.1.9 -- I'll see if we can get a repro there -- unfortunately, it may be delayed.

**Comment by rtestardi** [comment]
Steve and I confirm the bug still exists in 3.1.10.

**Comment by Jozkee** [comment]
Just as an update, I tried to repro the issue using [WinNFSd](https://github.com/winnfsd/winnfsd) on a Windows Pro machine and then mounting to the nfs server with WSL2 Ubuntu… I had to use NFSv3 since v4 is not supported on WinNFSd. 

However I wasn't able to repro the issue with said config; I ran the provided code on the linux client and it successfully writes to the file in the nfs server. 
This is how I am mounting it:
```
sudo mount -t nfs -o nfsvers=3 192.168.0.2:/c/share /mnt/share
```

My .NET version is 3.1.405.

I wonder if this issue is exclusive to NFSv4 and/or Windows Server's official NFS support.

I'll keep digging.

**Comment by Jozkee** [comment]
@rtestardi instead of using WriteAllText, which by default requests a shared lock, can you please try to create the FileStream yourself requiring an exclusive lock in the .ctor and then write to it? Use below snippet:
```cs
using System;
using System.IO;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            if( args.Length != 1 )
            {
                    Console.WriteLine("Usage: test <folder>\n");
                    return;
            }

            var filename = Path.Combine(args[0], "file.txt");
            // File.WriteAllText(filename, $"Hello World as of {DateTime.Now}");
            FileStream fs = File.Open(filename, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
            using (var sw = new StreamWriter(fs))
            {
                sw.Write($"Hello World as of {DateTime.Now}".AsSpan());
            }
            Console.WriteLine($"File Content: '{File.ReadAllText(filename)}'");
        }
    }
}
```

**Comment by danmoseley** [comment]
Pasting in offline notes:

For the second problem, it sounds like 28099 which definitely went out in 3.1.9 (and in 5.0). You reported the problem on 3.1.10, so it needs further investigation. Based on the description, that should be fixed if it’s not.

For the first problem, to summarize my understanding: StreamWriter (which File.WriteAllText uses) defaults to FileShare.Read (as does FileStream). This has been the case a long time, certainly back in .NET Framework, so it was designed for Windows API, and in Windows I guess it’s fairly typical to allow FileShare.Read (ie FILE_SHARE_READ) when you plan to write (FileAccess.Write, GENERIC_WRITE). Later when FileStream was ported to Unix, where locking is only advisory, we figured FILE_SHARE_READ mapped best to LOCK_SH:

corefx/FileStream.Unix.cs at f168bc36b744e630d89a34b7c90b603458e0e679 · dotnet/corefx (github.com)
            // Lock the file if requested via FileShare.  This is only advisory locking. FileShare.None implies an exclusive 
            // lock on the file and all other modes use a shared lock.  While this is not as granular as Windows, not mandatory, 
            // and not atomic with file opening, it's better than nothing.
            Interop.Sys.LockOperations lockOperation = (share == FileShare.None) ? Interop.Sys.LockOperations.LOCK_EX : Interop.Sys.LockOperations.LOCK_SH;
            if (Interop.Sys.FLock(_fileHandle, lockOperation | Interop.Sys.LockOperations.LOCK_NB) < 0)

which would be fine except that you report that the Windows NFS server implementation rejects writes under LOCK_SH.

1.	I’m guessing if we changed FileStream to map FileShare.Read to LOCK_EX it might be quite breaking at this point – any thoughts?

2.	Do I understand correctly that you believe Linux NFS servers allow writes under LOCK_SH – if so, is Windows intentionally behaving differently? Wouldn’t that imply any client of a Linux NFS server could break against a Windows NFS server?

3.	Is there an easy way (eg docker container) to set up such a Windows NFS server?

**Comment by Jozkee** [comment]
> I’m guessing if we changed FileStream to map FileShare.Read to LOCK_EX it might be quite breaking at this point – any thoughts?

Agree; I don't think that's something that we should change if this the only backing scenario that proves that FIleShare.Read = LOCK_EX. IMO it would require more investigation. cc. @carlossanlop @adamsitnik.

**Comment by rtestardi** [comment]
But our customers are using WriteAllText...

We're not doing this, they are -- we can tell them not to use C#, but they (hopefully) won't listen...

And if they use C#, they get a silent failure on a trivial one line program and no data is written, and they come to us asking where their data is! :-)

It seems .NET is losing the data, so .NET should be fixed?

**Comment by rtestardi** [comment]
Not sure I agree with that analysis...

If the handle was opened for write intent (not the same as the sharing mode), then why would lock(SH) ever be appropriate?

The lock is advisory on Linux, so it doesn't really matter -- but on Windows, it is enforced so the current code is guaranteed to fail, 100% of the time against Windows NFS server...

I'm pretty sure we can do the right (write :-) thing here and not break anything.

**Comment by adamsitnik** [comment]
> The following native C program exhibits the same issue

I am not sure if I understand. What `close()` method returns in the C program? A success or a failure?

> I’m guessing if we changed FileStream to map FileShare.Read to LOCK_EX it might be quite breaking at this point – any thoughts?

@danmoseley  I think that it would be just enough to change all the `File.Write*` APIs:

```cs
public static partial class File
{
    public static void WriteAllBytes(string path, byte[] bytes) { }
    public static Task WriteAllBytesAsync(string path, byte[] bytes, CancellationToken cancellationToken) 
    public static void WriteAllLines(string path, IEnumerable<string> contents) { }
    public static void WriteAllLines(string path, IEnumerable<string> contents, System.Text.Encoding encoding) { }
    public static void WriteAllLines(string path, string[] contents) { }
    public static void WriteAllLines(string path, string[] contents, System.Text.Encoding encoding) { }
    public static Task WriteAllLinesAsync(string path, IEnumerable<string> contents, System.Text.Encoding encoding, CancellationToken cancellationToken) 
    public static Task WriteAllLinesAsync(string path, IEnumerable<string> contents, CancellationToken cancellationToken) 
    public static void WriteAllText(string path, string? contents) { }
    public static void WriteAllText(string path, string? contents, System.Text.Encoding encoding) { }
    public static Task WriteAllTextAsync(string path, string? contents, Encoding encoding, CancellationToken cancellationToken) 
    public static Task WriteAllTextAsync(string path, string? contents, CancellationToken cancellationToken) 
}
```

to use `FileShare.None`, instead of using default `StreamWriter` ctor that uses `FileShare.Read` which is translated to the shared lock:

https://github.com/dotnet/runtime/blob/9cfe6966b4b6cae834a7fb58b71cdd9097381330/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L168

https://github.com/dotnet/runtime/blob/9cfe6966b4b6cae834a7fb58b71cdd9097381330/src/libraries/System.Private.CoreLib/src/System/IO/LegacyFileStreamStrategy.Unix.cs#L58-L59

BTW this is what `File.OpenWrite` already does so I assume this would be acceptable for `File.Write*` methods:

https://github.com/dotnet/runtime/blob/9cfe6966b4b6cae834a7fb58b71cdd9097381330/src/libraries/System.IO.FileSystem/src/System/IO/File.cs#L263-L267

@rtestardi Even with good intentions in mind, this would be a breaking change and we don't backport breaking changes to LTS releases. And even if we would hypothetically convince the shiproom to do it, your customers would still need to update their .NET Core installation. Not to mention how much time it would take.

> we can tell them not to use C#, but they (hopefully) won't listen

Instead of this, we would like to ask you to ask them to use a simple workaround:

```cs
public static void WriteAllText(string path, string contents)
{
    using (FileStream fileStream = new FileStream(path, FileMode.Create, FileAccess.Write, FileShare.None /* LOCK_EX */))
    using (StreamWriter sw = new StreamWriter(fileStream))
    {
        sw.Write(contents);
    }
}
```

I hope that copying eight lines of code and getting unblocked immediately is a good trade-off. In the meantime, we are going to revisit this approach and hopefully change it for .NET 6 (our next major release).

**Comment by danmoseley** [comment]
cc @stephentoub

**Comment by rtestardi** [comment]
> Instead of this, we would like to ask you to ask them to use a simple workaround:
The problem is the customer can't get past "the program does not work" without extensive research.

No API returns an error (if it did, and the customer noticed, surely we could point them to a workaround based on that!).

Instead, what happens is the customer tells us the data is silently lost, and eventually simply decides either a) don't use .NET, or b) don't use Windows NFS Server -- they never figure out why...

That's bad for us, of course, either way. :-(

I probably don't understand why this has to be a "breaking change" -- wouldn't it actually be a "fixing change" -- the worst that can happen is a silent failure turns into either a non-silent error (good) or a silent success (even better)?

What breaks that used to work, exactly?

> I hope that copying eight lines of code and getting unblocked immediately is a good trade-off.

The customer will never get to the point where they can identify a write buried beep in libraries (that they may not even control) that claims to succeed, while actually failing and doing nothing, unfortunately.

They will simply go elsewhere. :-(

How is this a good trade-off?

**Comment by Jozkee** [comment]
> What breaks that used to work, exactly?

@rtestardi Programs that were opening files with shared locks will now be opening them with exclusive ones, which will cause that other programs react different when opening the same file.

Take this code for example:
```cs
var fs1 = new FileStream("foo.txt", FileMode.OpenOrCreate, FileAccess.Write, FileShare.Write);
var fs2 = new FileStream("foo.txt", FileMode.OpenOrCreate, FileAccess.Write, FileShare.Write);

Console.WriteLine("Finished");
```

As of today, it is completely fine to have two FileStreams opening the same file as long as they don't use `FileShare.None`. 

If we decide to change the locking behavior to what you suggest, above code would throw "The process cannot access the file 'foo.txt' because it is being used by another process."

**Comment by stephentoub** [comment]
This functionality exists in the first place as an attempt to emulate (as best we could muster) the sharing behavior that exists on Windows and that .NET code was then expecting on Linux.  Developers were relying on FileStreams created with FileShare.None preventing other concurrent usage of that file with another FileStream.  So given the limited capabilities available with POSIX to emulate FileShare behaviors, we mapped FileShare.None to an exclusive lock and everything else to a shared lock, such that opening a file with FileShare.None would prevent concurrent usage with FileShare.Read/Write, but concurrent usage with anything other than None would be allowed.  And we know developers are actually expecting said exceptions and relying on them, because we've received multiple issues about it not working correctly, either not respecting even more of the meaning of FileShare (e.g. differentiating Read and Write) or when we accidentally broke this behavior in a previous release and got issues filed that code which was expecting IOExceptions was no longer getting them and concurrency that should have been prohibited was silently allowed.

From my perspective, we have a few options:
1. Make no changes in FileStream, and work with the powers that be around Windows NFS to improve behaviors here when accessing Windows NFS from Linux, e.g. ensuring that writes that fail actually return an error code indicating the failure.
2. Rip out all of this FileShare logic entirely, making it a nop on Linux.  This is a breaking change, in that code relying on FileShare.None vs anything else for protection against concurrent usage with other FileStreams will now be silently broken.
3. Enable opting-out (or as a variant of (2) opting in) to this behavior via an environment variable or other form of switch, e.g. if a particular environment variable is set, FileStream doesn't do any locking.  An opt-out would provide a no-code-change workaround if someone knows they're hitting the issue.  An opt-in would mean someone being silently broken due to unexpected concurrent usage would have a workaround when discovered.
4. Find some way of detecting in FileStream on Linux that it's accessing a Windows NFS server and don't do the advisory locking in such cases.
5. Some limited variant of (2) where on Linux we use FileShare.None in more cases to limit the impact of this with Windows NFS Servers, while also incurring some of the breaks mentioned.
6. Find some other heretofore unknown way of implementing FileShare in FileStream on Linux that doesn't rely on file locks.

**Comment by rtestardi** [comment]
Unfortunately, this is 100x easier to hit than I thought -- no need to write C# -- just use powershell from Linux:

```
PS /home/storsimple> "1" | set-content /mnt/richardtshare1/data1/ping.txt
PS /home/storsimple> dir /mnt/richardtshare1/data1/ping.txt

    Directory: /mnt/richardtshare1/data1

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-----           3/29/2021  8:07 PM              0 ping.txt

PS /home/storsimple>
```

I don't get how we can think this is anything but a critical .NET bug. :-(

**Comment by stephentoub** [comment]
@rtestardi, you said/asked offline:
"With that said, if I look at the traces, flock(SH) followed by write() just seems philosophically wrong. What benefit does the flock(SH) give for writes -- can we just get rid of it and make everyone happy (and faster!)?"

I thought I answered this above in https://github.com/dotnet/runtime/issues/44546#issuecomment-786270267, but I'll rephrase.

I think this question stems from perspective viewing flock(SH) as being equivalent to a “read” lock specific to performing file reads.  But in the general case on Linux, it’s equivalent to a shared lock, not specific to reads (noting the man page just says "Place a shared lock" rather than saying it's specific to reads).  It just means any number of callers can have access as long as no exclusive user has access, regardless of whether those shared or exclusive users are performing reads or performing writes.  Lacking any better mechanism for achieving it, .NET on Linux has implemented limited support for FileShare then via this locking mechanism, so that FileStreams created with FileShare.None requests exclusive access and everything else requests shared access, which enables FileStreams created with FileShare.None to know that no other FileStream will be concurrently accessing the same file.  So to answer your question, that is the benefit this whole thing provides: providing basic support for concurrent/exclusive access, and getting rid of it means we will break code that depends on that.

Of course, as this issue highlights, things break down with NFS because flocks may end up being translated such that flock(SH) is actually read-specific, and so in that light, it is antithetical to use flock(SH) with writes.

The options as I see them remain the same as I outlined a month ago in https://github.com/dotnet/runtime/issues/44546#issuecomment-786270267.  You're suggesting what I outlined as (2).

**Comment by stwolfe** [comment]
The fundamental problem here is the slight lose of data.  If we could just get an error in this case wouldn't that be an adequate fix?   It would not be a breaking change, because today it is slight data loss ( absolutely NO good ), and tomorrow would be an error...something that can be dealt with.

**Comment by rtestardi** [comment]
Actually, since writes can fail for reasons other than flock(SH), the part of the bug Steve refers to needs to be fixed regardless...  Simply failing the doomed write would be enough to satisfy me (then folks could look up the workaround to use -o nolock?).  But I don't know where that failure status is lost -- somewhere between NFS Server, NFS Client, Linux buffer cache, and .NET...

> You're suggesting what I outlined as (2)

I think I'm suggesting less than that...  How about simply don't do flock(SH) or do flock(EX) on handles that have write permission?  Handles without write permission need not change.

Again, flock(SH) followed by write() just seems philosophically wrong to me -- what benefit does it offer?  (Yes, I understand folks might open with write permission even when they have no intention of actually writing.)

**Comment by stephentoub** [comment]
> Again, flock(SH) followed by write() just seems philosophically wrong to me -- what benefit does it offer?

I've answered this in both of my last two responses.  We're obviously talking last each other.  Can you help me understand which part of my answer isn't clear so that I may try again?

**Comment by rtestardi** [comment]
I've read your answers -- and it still seems wrong to me. :-)  I'm guessing we just see things differently.

**Comment by stephentoub** [comment]
So when you wrote "Again, flock(SH) followed by write() just seems philosophically wrong to me -- what benefit does it offer?"... you understood what I outlined as the benefits but didn't agree they were benefits?

**Comment by rtestardi** [comment]
I'm certain you can come up with a scenario with concurrent readers and writers that sounds beneficial.  I just don't think that can justify the current "single writer" behavior -- or the current ambiguities.

I understand you want to decide how to flock() based on FileShare, but I think you should consider FileAccess *in addition* to FileShare.

**Comment by stephentoub** [comment]
> I'm certain you can come up with a scenario with concurrent readers and writers that sounds beneficial

Yes, not just sounding beneficial, there are scenarios / customer code exists that is explicitly about using a file as the hand off for data between writers and readers.

> I understand you want to decide how to flock() based on FileShare, but I think you should consider FileAccess in addition to FileShare.

FileAccess doesn't factor into the sharing decision on Windows, and the whole point of this was to approximate as best as possible the Windows behavior which in turn was the long-expected .NET behavior.  Changing it to suddenly factor in FileAccess would be much more breaking than just not respecting it at all.

**Comment by rtestardi** [comment]
Sure, but even the "handoff" scenario can't outweigh this, right?

```
PS /home/storsimple> "1" | set-content /mnt/richardtshare1/data1/ping.txt
PS /home/storsimple> dir /mnt/richardtshare1/data1/ping.txt

    Directory: /mnt/richardtshare1/data1

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-----           3/29/2021  8:07 PM              0 ping.txt

PS /home/storsimple>
```

How do we fix the simple, single-threaded, single-writer write?

**Comment by stephentoub** [comment]
> Sure, but even the "handoff" scenario can't outweigh this, right?

As I noted earlier, the more important scenario is folks creating FileShare.None FileStreams, expecting other attempts to create FileStreams to fail.  If we, for example, delete the FileShare-related code, such concurrency prevention will simply silently go away.

> How do we fix the simple, single-threaded, single-writer write?

I outlined all of the options I'm aware of.

**Comment by rtestardi** [comment]
I'd hardly say returning an error on concurrent access is "more important" than not silently corrupting data for most basic non-concurrent access, but we can disagree there...

Can I ask why not just do as I suggested from the opening post  -- use FileShare *and* FileAccess to decide the flock() mode?

```
if (FileShare.None || FileAccess.Write) {
    use flock(EX)
} else {
    use flock(SH)
}
```
What is wrong with that?  It seems to cover your "more important" scenario as well as the corruption of data on the most basic non-concurrent access.

Have I overlooked something?  All the behavioral changes are "safe" -- you might get an error where you used to get a success (in undocumented ambiguously mapped behavior), but you no longer silently corrupt data.

**Comment by rtestardi** [comment]
Let's call my suggestion #7 or #0, to differentiate it from your #1 - #6.

**Comment by stephentoub** [comment]
> Can I ask why not just do as I suggested from the opening post -- use FileShare and FileAccess to decide the flock() mode? What is wrong with that? 

That's what I just answered, and what you answered as well: "you might get an error where you used to get a success".

> Let's call my suggestion 7

Ok.

**Comment by danmoseley** [comment]
Leaving the locking aside, should we try to understand where the error is lost?

**Comment by rtestardi** [comment]
IIRC, forcing write thru (O_SYNC or -o sync?) on the handle may make the error propagate correctly -- which would be more evidence that an fsync() might need to be added to .NET before close().  Someone would need to verify this.

**Comment by filipnavara** [comment]
It would be nice to reach some consensus on the file sharing emulation. I started rebuilding my `File.Copy` PR and this was one of the reasons why it was reverted last time.

**Comment by jeffhandley** [comment]
We haven't yet been able to get back to this issue to decide on the approach we'd want to take to work around this underlying behavior. It's unlikely that we will be able to get this resolved in time for 6.0 Preview 7, and I don't expect we'd consider it an RC1 candidate; so this is likely going to slip to 7.0.

**Comment by filipnavara** [comment]
At this point I expected this to slip into the .NET 7 timeframe. I'll keep this on my radar and come back to ask later.
